department-of-veterans-affairs/caseflow
# hotfix/Sort Veteran Participant ID Column Numerically

Resolves [APPEALS-14301](https://vajira.max.gov/browse/APPEALS-14301)

### Description

The Veteran Participant ID column's values are strings of numbers. It was noticed that these values are being sorted using string comparisons instead of numeric comparisons, which causes tasks to appear out of order.

With string comparison, strings are compared character-by-character, from left-to-right. This means that the string `"6231"` will be considered to be greater than `"60000000000000000001"` since the second character of `"6231"` (`2`) has a higher ASCII value than that of `"60000000000000000001"` (`0`).

This pull request casts the values of the participant ID column to integers before ordering them so that `60000000000000000001` can be identified as a larger value than `6231`.

### Testing Plan
1. Run the following commands in the Rails console in order to populate the VHA business line's decision review queue:
```ruby
new_vha_tasks = Array.new(100).map { FactoryBot.create(:higher_level_review_vha_task) }

Array.new(100).each { FactoryBot.create(:supplemental_claim_vha_task) }

new_vha_tasks.each do |task|
  task.appeal.veteran.update!(participant_id: rand(100..999_999_999).to_s)
end
```
2. Log into Caseflow as a VHA business line user (ex: `ACBAUERVVHAH`)
3. Navigate to the VHA's decision review queue ([Link for locahost](http://localhost:3000/decision_reviews/vha))

![no-sort](https://user-images.githubusercontent.com/99351305/213898680-ee427fc3-a0ec-41da-9e1f-92eecb74ee0d.png)

4. Click on the "double-arrow" icon to the right of the "Veteran Participant Id" column header.
    - The URL should now be `/decision_reviews/vha?tab=in_progress&page=1&sort_by=veteranParticipantIdColumn&order=desc`
    - The first row on the first page should contain the highest participant ID (numerically).
    - As your move through subsequent pages, the participant IDs should **decrease**.
    - The last row on the last page should contain the lowest participant ID (numerically).
  
![desc-sort](https://user-images.githubusercontent.com/99351305/213898684-589a4bdd-dc0b-4587-8f67-c92b0bb8e644.png)

5. Click on the "double-arrow" icon again.
    - The URL should now be `/decision_reviews/vha?tab=in_progress&page=1&sort_by=veteranParticipantIdColumn&order=asc`
    - The first row on the first page should contain the lowest participant ID (numerically).
    - As your move through subsequent pages, the participant IDs should **increase**.
    - The last row on the last page should contain the highest participant ID (numerically).
  
![asc-sort](https://user-images.githubusercontent.com/99351305/213898689-604559de-56fe-4178-b2a9-1a1a4d97eac7.png)

6. Enter a search query that corresponds with some subset of the queue. Validate that the sorting for the participant ID column works with a search query active.
7. Feel free to clear the search query at this time.
8. Filter the queue by task type by clicking on the "funnel" icon next to the "Type" column header. Select either SC or HLR, but not both. Validate that the sorting for the participant ID column works with a task filter active.
9. You may clear the filter by clicking on the "Clear all filters" link above the queue table.
10. Please ensure that the sorting behaviors of all other columns have been left intact.
11. With a column sort active, refresh the page. Ensure that the column sort you've select (column and direction) are still active after the page reloads.
